### ISSUE TITLE
docs(notebooks): regression masterclass — pattern 1 (eqx.tree_at + raw NumPyro)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#75]
## Problem / Request

Port the [Regression Masterclass design doc](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) to an executed `docs/notebooks/regression_masterclass_treeat.ipynb`, presenting **the same 7 regression models** built with the rawest possible pattern: bare `eqx.Module`s with placeholder weights, `numpyro.sample` calls in the model function, and `eqx.tree_at` as the bridge that injects sampled values into the frozen module. **No `pyrox`-specific abstractions** — this notebook teaches the canonical Equinox + NumPyro + einops pattern from scratch.

This is the **first of three sibling notebooks** that solve the same `f(x) = sin(x)/x + ε` regression task with progressively higher-level pyrox abstractions:

1. **This issue** — `tree_at` (raw equinox + numpyro)
2. Sibling — [`pyrox_sample` (PyroxModule)](#)
3. Sibling — [`Parameterized` (PyroxParam declarative)](#)

The three notebooks land together so users can read them side-by-side and watch the boilerplate shrink.

## User Story

As a JAX user new to probabilistic deep learning, I want to read one notebook that builds **seven** Bayesian regression models — Bayesian linear, MAP MLP, MC-Dropout MLP, full Bayesian MLP, RFF SVR, hierarchical RFF GP, and deep RFF GP — using nothing but `equinox` modules, `numpyro.sample` calls, and `eqx.tree_at`. I want every contraction written as `einsum` with named axes so I can read the shape semantics at a glance. I want to come away knowing exactly how to bridge a frozen Equinox PyTree to a NumPyro probabilistic program.

## Motivation

1. **Pedagogy first.** The pyrox abstractions (`PyroxModule`, `Parameterized`) are leaky if users never see what they replace. Showing the raw pattern first justifies every line of the higher-level patterns in the sibling notebooks.
2. **Self-contained adoption.** A reader who only ever uses `equinox` + `numpyro` should still be able to copy this notebook and ship a regression model. The notebook does not require `pyrox` at all in its code cells (only as the host package providing the docs site).
3. **Three-way comparison anchor.** Without the raw baseline, the sibling notebooks have nothing to compare against — the boilerplate-reduction story falls flat.

## Proposed Structure

Following [`.github/instructions/docs-examples.instructions.md`](.github/instructions/docs-examples.instructions.md):

1. **Title + Colab badge** — `# Regression Masterclass — Pattern 1: eqx.tree_at + raw NumPyro`.
2. **Background** — three-library architecture (Equinox / NumPyro / einops); the `eqx.tree_at` bridge pattern; why `einsum` over `@`.
3. **Setup** — Colab detection + `pip install -q "pyrox[colab] @ git+https://github.com/jejjohnson/pyrox@main"` (pulls in numpyro, equinox, einops as transitive deps).
4. **Imports + `jax_enable_x64` + watermark**.
5. **Dataset** — `f(x) = sin(x)/x + ε`, 100 training points in `[-10, 10]`, dense test grid in `[-20, 20]`. **Inlined** — no helper functions.
6. **Model 1 — Bayesian linear (polynomial features)** — `LinearRegressor(eqx.Module)` + `model_linear` + NUTS.
7. **Model 2 — MLP via MAP** — `MLP(eqx.Module)` + `model_nnet` + SVI with `AutoDelta`.
8. **Model 3 — MC-Dropout MLP** — `MLPDropout(eqx.Module)` with explicit mask as `numpyro.sample` site.
9. **Model 4 — Bayesian MLP via NUTS** — same `MLP` module, full posterior.
10. **Model 5 — SVR via RFF** — `RFFFeatureMap(eqx.Module)` with fixed frequencies; only weights inferred.
11. **Model 6 — Approximate GP via hierarchical RFF** — same RFF, hierarchical amplitude prior.
12. **Model 7 — Deep RFF GP** — `DeepRFFRegressor(eqx.Module)` composing two RFF layers.
13. **Side-by-side summary table** — model / module / `einsum` patterns / inference method / R² on held-out.
14. **Pointer to siblings** — "see `regression_masterclass_pyrox_sample.ipynb` for the same models with `PyroxModule`, and `regression_masterclass_parameterized.ipynb` for the declarative `Parameterized` form."

**No helpers.** The original design doc shipped `make_dataset`, `run_mcmc`, `run_svi`, `predict`, `summarise_predictions`, `compute_r2` as module-level functions. The notebook **inlines all of them**:

| Helper | Inlined replacement |
|---|---|
| `make_dataset(...)` | 5-line `jr.uniform` + `jnp.sinc` + `jr.normal` block |
| `run_mcmc(...)` | `MCMC(NUTS(model_fn), num_warmup=..., num_samples=...).run(rng, x_train, y_train, **kwargs); samples = mcmc.get_samples()` |
| `run_svi(...)` | `SVI(model_fn, guide, optax.adam(1e-3), Trace_ELBO()).run(rng, num_steps, x_train, y_train, **kwargs)` |
| `predict(...)` | `Predictive(model_fn, posterior_samples=samples)(rng, x_test, **kwargs)["obs"]` |
| `summarise_predictions(...)` | `mean = preds.mean(0); lo, hi = jnp.quantile(preds, jnp.array([0.025, 0.975]), axis=0)` |
| `compute_r2(...)` | `1 - jnp.sum((y - yh)**2) / jnp.sum((y - y.mean())**2)` |

## Mathematical Notes

The math for all 7 models is documented in full in [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.md`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md). The notebook should reproduce the math blocks from that doc verbatim. Key takeaways the reader should leave with:

### The bridge

Equinox modules are frozen PyTrees: `net.weight = w` is impossible. Instead,

$$\text{net}_{\text{posterior}} \;=\; \texttt{eqx.tree\_at}\bigl(\lambda m\!:\!m.\text{weight},\, \text{net}_{\text{init}},\, w_{\text{sampled}}\bigr)$$

returns a *new* module identical to the old except at the selected leaf. For multi-leaf injection use a tuple selector `lambda m: (m.W1, m.b1, m.W2, m.b2)` with a tuple of replacements.

### MAP via `AutoDelta` is exact

Maximizing the ELBO with a delta guide is equivalent to maximizing the log-joint:

$$\theta_{\text{MAP}} \;=\; \arg\max_\theta \log p(\theta \mid \mathcal{D}) \;=\; \arg\max_\theta \bigl[\log p(\mathcal{D} \mid \theta) + \log p(\theta)\bigr].$$

So `SVI(model_fn, AutoDelta(model_fn), ..., Trace_ELBO())` *is* MAP.

### MC-Dropout as a numpyro sample site

Making the dropout mask an explicit `numpyro.sample("dropout_mask", dist.Bernoulli(...))` means `Predictive` automatically draws fresh masks during prediction:

$$z_{ij} \sim \text{Bernoulli}(1 - p_{\text{drop}}), \qquad \tilde h_i = h_i \odot z_i / (1 - p_{\text{drop}})$$

### RFF as a Bochner-theorem approximation

For the RBF kernel $k(x, x') = \exp\!\bigl(-\lVert x - x'\rVert^2 / (2\ell^2)\bigr)$:

$$\omega_j \sim \mathcal{N}(0, \ell^{-2}), \quad b_j \sim \text{Uniform}(0, 2\pi), \qquad \phi(x)_j = \sqrt{2/D}\,\cos(\omega_j x + b_j)$$

gives $\phi(x)^\top \phi(x') \approx k(x, x')$. Inference becomes Bayesian linear regression in $D$-dimensional feature space — $O(D^3)$ instead of $O(N^3)$.

## Reference Code

The full reference implementation already exists at [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.py`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.py) (2000 LOC, 7 model classes + 7 numpyro models + 4 inference helpers). The notebook ports that file's *model definitions and numpyro models verbatim* and **inlines** the helper functions at each call site instead of importing them.

Pattern translation example — `LinearRegressor`:

```python
# Module — pure equinox, placeholder weights
class LinearRegressor(eqx.Module):
    weight: jax.Array
    degree: int = eqx.field(static=True)

    def __init__(self, degree: int = 1):
        self.degree = degree
        self.weight = jnp.zeros(degree + 1)

    def features(self, x):
        return jnp.stack([x ** p for p in range(self.degree + 1)], axis=-1)

    def __call__(self, x):
        return einsum(self.features(x), self.weight, "n feat, feat -> n")


# NumPyro model — sample, then inject via tree_at
def model_linear(x, y=None, *, degree=1):
    net = LinearRegressor(degree=degree)
    w = numpyro.sample("w", dist.Normal(jnp.zeros(degree + 1), 1.0))
    net = eqx.tree_at(lambda m: m.weight, net, w)
    f = numpyro.deterministic("f", net(x))
    sigma = numpyro.sample("sigma", dist.HalfNormal(1.0))
    numpyro.sample("obs", dist.Normal(f, sigma), obs=y)
```

The four-leaf MLP variant uses a tuple selector (see [`design_docs/...regression_masterclass_eqx.py:982-1063`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.py)):

```python
net = eqx.tree_at(lambda m: (m.W1, m.b1, m.W2, m.b2), net, (W1, b1, W2, b2))
```

## References

1. Aboleth tutorial — [Regression Master Class](https://aboleth.readthedocs.io/en/stable/tutorials/some_regressors.html). The original TensorFlow tutorial this port reimplements.
2. Kidger, P. (2021). *equinox: neural networks in JAX via callable PyTrees and filtered transformations*. [Docs](https://docs.kidger.site/equinox/) — `eqx.Module`, `eqx.tree_at`, the immutable-pytree pattern.
3. Phan, D., Pradhan, N., & Jankowiak, M. (2019). *Composable Effects for Flexible and Accelerated Probabilistic Programming in NumPyro*. [arXiv:1912.11554](https://arxiv.org/abs/1912.11554) — `numpyro.sample`, `numpyro.deterministic`, the trace-based programming model.
4. Rahimi, A. & Recht, B. (2007). *Random Features for Large-Scale Kernel Machines*. NeurIPS. [PDF](https://people.eecs.berkeley.edu/~brecht/papers/07.rah.rec.nips.pdf) — RFF / Bochner theorem (Models 5–7).
5. Cutajar, K., Bonilla, E. V., Michiardi, P., & Filippone, M. (2017). *Random Feature Expansions for Deep Gaussian Processes*. ICML. [arXiv:1610.04386](https://arxiv.org/abs/1610.04386) — Deep RFF (Model 7).
6. Gal, Y. & Ghahramani, Z. (2016). *Dropout as a Bayesian Approximation*. ICML. [arXiv:1506.02142](https://arxiv.org/abs/1506.02142) — MC-Dropout (Model 3).
7. Hoffman, M. D. & Gelman, A. (2014). *The No-U-Turn Sampler*. JMLR 15. [arXiv:1111.4246](https://arxiv.org/abs/1111.4246) — NUTS, the MCMC engine used throughout.
8. Rogozhnikov, A. (2022). *einops: Clear and Reliable Tensor Manipulations with Einstein-like Notation*. ICLR. [arXiv:2106.04788](https://arxiv.org/abs/2106.04788) — einsum-based contractions.
9. [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.md`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) and [`.py`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.py) — the reference implementation this notebook ports.

## Implementation Steps

- [ ] Author `docs/notebooks/regression_masterclass_treeat.py` in jupytext percent format.
- [ ] Header + first markdown cell: title + Colab badge to `jejjohnson/pyrox@main`.
- [ ] Setup cell: Colab detection + `pip install pyrox[colab]` (transitively installs numpyro, equinox, einops).
- [ ] Imports + `jax.config.update("jax_enable_x64", True)` + `%watermark -p jax,equinox,numpyro,einops`.
- [ ] Dataset cell — inline 5-line `make_dataset` body.
- [ ] One section per model (1–7): markdown-math block, `eqx.Module` definition, `numpyro` model fn, inference call (inline `MCMC(NUTS(...))` or `SVI(..., AutoDelta(...), ...)`), prediction (inline `Predictive(...)`), R² + 95% interval plot.
- [ ] Closing summary table + sibling-notebook pointer.
- [ ] Convert: `uv run --group docs jupytext --to notebook docs/notebooks/regression_masterclass_treeat.py`.
- [ ] Execute: `uv run --group docs jupyter nbconvert --to notebook --execute docs/notebooks/regression_masterclass_treeat.ipynb --inplace --ExecutePreprocessor.timeout=600`.
- [ ] Delete `.py`; commit executed `.ipynb`.
- [ ] `mkdocs.yml` nav entry under "Tutorials" (or new "Masterclass" section).

## Definition of Done

- `docs/notebooks/regression_masterclass_treeat.ipynb` is committed with embedded matplotlib outputs for all 7 models.
- Notebook executes end-to-end under `nbconvert --execute --inplace --timeout=600` from a clean `uv sync --all-groups` environment.
- Each model's R² on held-out is asserted in a cell:
  - Linear (degree=1): R² ≈ 0 (sanity — line can't fit sinc; prove the assertion harness works)
  - MAP MLP: R² > 0.85
  - MC-Dropout MLP: R² > 0.85
  - Bayesian MLP (NUTS): R² > 0.90
  - SVR/Approx GP/Deep GP via RFF: R² > 0.85
- The notebook imports **zero** `pyrox.*` symbols inside code cells (it's purely `equinox` + `numpyro` + `einops`).
- The notebook contains **no** `from <helpers> import ...` lines — every helper is inlined at the call site.

## Testing

- The notebook's `assert R² > X` cells are the primary regression tests.
- Add the notebook to a docs-smoke CI job (if one exists) that runs `nbconvert --execute --inplace` on all `docs/notebooks/*.ipynb` to catch upstream `numpyro` / `equinox` API drift.

## Documentation

- This *is* the docs deliverable. The notebook is the doc.
- Cross-link to it from `docs/index.md` under a new "Tutorials → Regression Masterclass" group.

## Relationships

- **Parent epic:** [#36 — `[Epic] 4.C NN Uncertainty Patterns`](https://github.com/jejjohnson/pyrox/issues/36).
- **Sibling notebooks** (filed together):
  - Pattern 2 — same models with `PyroxModule` + `pyrox_sample`.
  - Pattern 3 — same models with `Parameterized` + `PyroxParam` + native `pyrox.gp` for Models 6–7.
- **Reference doc:** [`design_docs/pyrox/examples/nn/regression_masterclass_eqx.md`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.md) + [`.py`](design_docs/pyrox/examples/nn/regression_masterclass_eqx.py).
- **Not blocked by** any other open issue. Backbone (`equinox`, `numpyro`, `einops`) is fully present today.
